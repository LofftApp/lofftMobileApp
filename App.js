import React, {useState, useEffect} from 'react';
import Icon from 'react-native-vector-icons/Ionicons';
import Color from './src/styles/lofftColorPallet.json';

// FireStore ðŸ”¥
import auth from '@react-native-firebase/auth';
import firestore from '@react-native-firebase/firestore';

import SplashScreen from 'react-native-splash-screen';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';
import {navigationRef} from './src/navigation/RootNavigation';
import {createBottomTabNavigator} from '@react-navigation/bottom-tabs';
import SignUpScreen from './src/screens/SignUpScreen';
import SignInScreen from './src/screens/SignInScreen';
import {GoogleSignin} from '@react-native-google-signin/google-signin';

// Find Lofft Journey
import StartJourney from './src/screens/StartJourney';
import AboutYouFlatHuntScreen from './src/screens/renterJourneyScreens/AboutUserScreen';
import GenderIdentityScreen from './src/screens/renterJourneyScreens/GenderIdentityScreen';
import SelectCityScreen from './src/screens/renterJourneyScreens/SelectCityScreen';
import FinderBudgetScreen from './src/screens/renterJourneyScreens/FinderBudgetScreen';
import FlatFeaturesScreen from './src/screens/renterJourneyScreens/FlatFeaturesScreen';
import SelfDescribeScreen from './src/screens/renterJourneyScreens/SelfDescribeScreen';
import UserConditionsScreen from './src/screens/UserConditionsScreen';
import FlatListScreen from './src/screens/renterFlatFindScreens/FlatListScreen';

const Stack = createStackNavigator();
const Tab = createBottomTabNavigator();

const App = () => {
  // Set an initializing state whilst Firebase connects
  const [initializing, setInitializing] = useState();
  const [user, setUser] = useState(null);
  const [userType, setUserType] = useState(null);
  const [admin, setAdmin] = useState(false);

  const onAuthStateChanged = async user => {
    setUser(user);
    console.log(user);
    if (user) {
      const userClaims = await user.getIdTokenResult();
      setUserType(userClaims.claims.primaryUserType);
      setAdmin(userClaims.claims.admin);
    }
    if (initializing) setInitializing(false);
  };

  useEffect(() => {
    const subscriber = auth().onAuthStateChanged(onAuthStateChanged);
    return subscriber;
  }, []);

  GoogleSignin.configure({
    webClientId:
      '25055797109-13te2c0d3acitt9dvjs212ujgt4odr9q.apps.googleusercontent.com',
  });
  // Use Effect for dev environment
  useEffect(() => {
    if (__DEV__) {
      console.log('FireStore Development Environment');
      let host = 'localhost';
      // If using Mobile device set the host as local IP
      // host = '192.168.1.16';
      if (host === 'localhost') {
        console.log('Host running on local host');
      } else {
        console.log(`Host is running on ${host}`);
      }
      firestore().useEmulator(host, 8080);
      auth().useEmulator(`http://${host}:9099`);
    }
  }, []);

  if (initializing) return null;
  return (
    <>
      {user && userType ? (
        <Tab.Navigator
          screenOptions={({route}) => ({
            tabBarIcon: ({focused, color}) => {
              let iconName = 'Not sure';
              if (route.name === 'search') {
                iconName = 'search-outline';
              }
              return <Icon name={iconName} size={25} color={color} />;
            },
            tabBarActiveTintColor: Color.Lavendar[100],
            tabBarInActiveTintColor: Color.Black[30],
            tabBarShowLabel: false,
          })}>
          <Tab.Screen
            name="search"
            component={FlatListScreen}
            options={{headerShown: false}}
          />
        </Tab.Navigator>
      ) : (
        <Stack.Navigator initialRouteName="SignInScreen">
          {user ? (
            <>
              <Stack.Screen
                name="StartJourney"
                component={StartJourney}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="AboutYouFlatHuntScreen"
                component={AboutYouFlatHuntScreen}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="GenderIdentityScreen"
                component={GenderIdentityScreen}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="SelectCityScreen"
                component={SelectCityScreen}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="FinderBudgetScreen"
                component={FinderBudgetScreen}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="FlatFeaturesScreen"
                component={FlatFeaturesScreen}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="SelfDescribeScreen"
                component={SelfDescribeScreen}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="UserConditionsScreen"
                component={UserConditionsScreen}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="FlatListScreen"
                component={FlatListScreen}
                options={{headerShown: false}}
              />
            </>
          ) : (
            <>
              <Stack.Screen
                name="SignUpScreen"
                component={SignUpScreen}
                options={{headerShown: false}}
              />
              <Stack.Screen
                name="SignInScreen"
                component={SignInScreen}
                options={{headerShown: false}}
              />
            </>
          )}
        </Stack.Navigator>
      )}
    </>
  );
};

export default () => {
  return (
    <NavigationContainer
      ref={navigationRef}
      onReady={() => SplashScreen.hide()}>
      <App />
    </NavigationContainer>
  );
};
