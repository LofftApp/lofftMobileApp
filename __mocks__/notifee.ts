const notifee = {
  displayNotification: jest.fn(() => Promise.resolve()),
  createChannel: jest.fn(() => Promise.resolve('mock-channel-id')),
  onBackgroundEvent: jest.fn(() => Promise.resolve()),
};

export default notifee;
