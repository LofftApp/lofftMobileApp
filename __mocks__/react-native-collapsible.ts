const Collapsible = jest.mock('react-native-collapsible', () => {
  const actualCollapsible = jest.requireActual('react-native-collapsible');
  return {
    ...actualCollapsible,
    someOtherFunction: jest.fn(), // Mock specific functions if needed
  };
});

export default Collapsible;
