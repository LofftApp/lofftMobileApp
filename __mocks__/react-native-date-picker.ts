const DatePicker = {
  show: jest.fn(),
  hide: jest.fn(),
};

export default DatePicker;
