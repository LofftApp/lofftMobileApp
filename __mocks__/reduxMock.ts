import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

// Factory function to create a mock store with customizable state
export const createMockStore = (initialState = {}) => {
  const mockStore = configureStore([thunk]);
  return mockStore(initialState);
};

// Example of an initial state you can reuse in tests
export const defaultMockState = {
  authentication: {
    loading: false,
  },
  user: {
    user: {
      id: 1,
      email: 'james@example.com',
      admin: false,
      terms_accepted: false,
      user_type: 'tenant',
    },
  },
};
