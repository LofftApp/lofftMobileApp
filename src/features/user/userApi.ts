import {lofftApi} from 'reduxFeatures/api/lofftApi';
import {SpecificUser, User} from './types';
import {toCamelCaseKeys} from 'helpers/toCamelCaseKeys';
import { Platform } from 'react-native';
import {
  Assets,
  NewUserLessorDetails,
  NewUserTenantDetails,
} from 'reduxFeatures/registration/types';

export const userApi = lofftApi.injectEndpoints({
  endpoints: builder => ({
    getUser: builder.query<User, void>({
      query: () => 'api/users/profile',
      transformResponse: response => {
        console.log('getUser called ðŸ‘¾');
        return toCamelCaseKeys(response as User);
      },
      providesTags: [{type: 'User', id: 'PROFILE'}],
    }),
    getSpecificUser: builder.query<SpecificUser, number>({
      query: id => `api/users/${id}/specific_user`,
      transformResponse: response => {
        console.log('specific user called ðŸŽ‰');
        return toCamelCaseKeys(response as SpecificUser);
      },
    }),
    completeUserAndCreateTenant: builder.mutation<
      void,
      {id: number; userChoices: NewUserLessorDetails | NewUserTenantDetails; photos?: File[]}
    >({
      query: ({id, userChoices, photos}) => {
      const formData = new FormData();
      formData.append("userChoices", JSON.stringify(userChoices));
      if (photos) {
       photos.forEach((uri, index) => {
        formData.append(`photos[${index}]`, {
        uri: Platform.OS === 'ios' ? uri.toString().replace('file://', '') : uri, // iOS sometimes needs 'file://' removed
        name: `photo_${index}.jpg`,
        type: 'image/jpeg', // Adjust based on actual image format if needed
    });
  });
    }
      return {
        url: `/api/users/${id}/complete_tenant_sign_up`,
        method: 'POST',
        headers: {
          'Content-Type': 'multipart/form-data',
        },
        body: formData,
      };
    },
      invalidatesTags: [{type: 'User', id: 'PROFILE'}],
    }),

    getAssets: builder.query<Assets, void>({
      query: () => '/api/assets',
      transformResponse: response => {
        return toCamelCaseKeys(response as Assets);
      },
    }),
  }),
  overrideExisting: false,
});

export const {
  useGetUserQuery,
  useGetSpecificUserQuery,
  useCompleteUserAndCreateTenantMutation,
  useGetAssetsQuery,
} = userApi;
