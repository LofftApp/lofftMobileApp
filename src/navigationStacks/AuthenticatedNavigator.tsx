import React from 'react';
import {createNativeStackNavigator} from '@react-navigation/native-stack';

// Navigators ðŸ§­
import NewUserNavigator from 'navigationStacks/NewUserNavigator';
import TenantNavigator from 'navigationStacks/TenantNavigator';
import LessorNavigator from 'navigationStacks/LessorNavigator';
import AdminNavigator from './AdminNavigator';

//Types
import type {RootStackParamsList} from './types';

type AuthenticatedNavigatorProps = {
  userType?: string;
  admin?: boolean;
};

const AuthStack = createNativeStackNavigator<RootStackParamsList>();
const AuthenticatedNavigator = ({
  userType,
  admin,
}: AuthenticatedNavigatorProps) => {
  return (
    <AuthStack.Navigator screenOptions={{headerShown: false}}>
      {admin && <AuthStack.Screen name="AdminStack" component={AdminNavigator} />}
      {userType === 'newuser' && (
        <AuthStack.Screen name="NewUserStack" component={NewUserNavigator} />
      )}
      {userType === 'lessor' && (
        <AuthStack.Screen
          name="LessorDashboardStack"
          component={LessorNavigator}
        />
      )}
      {userType === 'tenant' && (
        <AuthStack.Screen
          name="TenantDashboardStack"
          component={TenantNavigator}
        />
      )}
    </AuthStack.Navigator>
  );
};

export default AuthenticatedNavigator;
