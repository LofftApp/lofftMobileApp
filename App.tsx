/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React, {useState, useEffect} from 'react';
import {Platform} from 'react-native';
import MapboxGL from '@rnmapbox/maps';
import {MAPBOX_API_KEY} from '@env';

// Redux 🏗️
import {useAppSelector, useAppDispatch} from 'reduxCore/hooks';
import {checkToken} from 'reduxFeatures/authentication/authenticationMiddleware';
import {createNativeStackNavigator} from '@react-navigation/native-stack';

import SplashScreen from 'react-native-splash-screen';
import {NavigationContainer} from '@react-navigation/native';
import {navigationRef} from './src/navigation/RootNavigation';
import {getProfile} from 'reduxFeatures/user/usersMiddleware';

// Navigators 🧭
import GuestStackNavigator from './navigationStacks/GuestNavigator';
import NewUserNavigator from './navigationStacks/NewUserNavigator';
import DashboardNavigator from './navigationStacks/DashboardNavigator';
// import LessorNavigator from './navigationStacks/LessorNavigator';
import DashboardNavigatorLessor from './navigationStacks/DashboardnavigtatorLessor';

// Dev Screesn 🛠️
import AdminScreen from 'screens/admin/adminScreen';
import {createSelector} from '@reduxjs/toolkit';

const RootStack = createNativeStackNavigator();

// Remove ErrorBoundary in production
import ErrorBoundary from './src/ErrorBoundary';

const App = () => {
  // Define selectors
  const getAuthenticated = (state: any) => state.authentication?.authenticated;
  const getUserType = (state: any) => state.user?.user?.userType;
  const getAdmin = (state: any) => state.user?.user?.admin;

  // Create memoized selectors
  const selectAuthenticated = createSelector(
    [getAuthenticated],
    authenticated => authenticated,
  );
  const selectUserTypeAndAdmin = createSelector(
    [getUserType, getAdmin],
    (userType, admin) => [userType, admin],
  );
  const authenticated = useAppSelector(selectAuthenticated);
  console.log('authenticated', authenticated);
  const [userType, admin] = useAppSelector(selectUserTypeAndAdmin);

  const dispatch = useAppDispatch();
  const [initializing, setInitializing] = useState(true);

  useEffect(() => {
    dispatch(checkToken());
  }, [dispatch]);

  useEffect(() => {
    if (initializing) {
      setInitializing(false);
    }
    if (authenticated && !userType) {
      dispatch(getProfile());
    }
  }, [authenticated, dispatch, initializing, userType]);

  // Mapbox
  MapboxGL.setWellKnownTileServer(
    Platform.OS === 'android' ? 'Mapbox' : 'mapbox',
  );
  MapboxGL.setAccessToken(MAPBOX_API_KEY);
  // This is needed to use Mapbox in offline mode and with android emulator
  MapboxGL.setTelemetryEnabled(false);

  // TODO: This will need to be placed in another useEffect with new DB path.

  // Use Effect for dev environment
  useEffect(() => {
    if (__DEV__) {
      console.log('Lofft API Development Environment');
      let host = 'localhost';
      // If using Mobile device set the host as local IP
      host = '127.0.0.1';
      console.log(
        host === 'localhost'
          ? 'Host running on local host'
          : `Host is running on ${host}`,
      );
    }
  }, []);
  return (
    <>
      {!authenticated ? (
        <GuestStackNavigator />
      ) : (
        <RootStack.Navigator screenOptions={{headerShown: false}}>
          {admin ? (
            <RootStack.Screen name="admin" component={AdminScreen} />
          ) : null}
          {!userType ? (
            <RootStack.Screen name="profileFlow" component={NewUserNavigator} />
          ) : null}
          {userType === 'lessor' ? (
            <RootStack.Screen
              name="dashboardLessor"
              component={DashboardNavigatorLessor}
            />
          ) : (
            <RootStack.Screen name="dashboard" component={DashboardNavigator} />
          )}
        </RootStack.Navigator>
      )}
    </>
  );
};

export default () => {
  return (
    <NavigationContainer
      ref={navigationRef}
      onReady={() => SplashScreen.hide()}>
      <ErrorBoundary>
        <App />
      </ErrorBoundary>
    </NavigationContainer>
  );
};
