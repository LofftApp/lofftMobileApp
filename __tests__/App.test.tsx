import 'react-native';
import React from 'react';
import App from '../App';
import {render} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {createMockStore} from '../__mocks__/reduxMock';
import {Store} from 'redux';
import AuthenticatedNavigator from 'navigationStacks/AuthenticatedNavigator';

// Mock lofftApi and injectEndpoints
jest.mock('../src/features/api/lofftApi', () => ({
  lofftApi: {},
}));

// Mock RTK Query hooks from userApi
jest.mock('../src/features/user/userApi', () => ({
  useGetUserQuery: jest.fn(() => ({})),
}));

// Mock RTK Query hooks from authApi
jest.mock('../src/features/auth/authApi', () => ({
  useSignOutMutation: jest.fn(() => [jest.fn(), {}]),
}));

// Mock the authSlice hook
jest.mock('../src/features/auth/authSlice', () => ({
  authSlice: jest.fn(),
}));

// Mock Navigators
jest.mock('../src/navigationStacks/GuestNavigator', () => {
  const {Text} = require('react-native');
  return () => <Text testID="guest-navigator" />;
});

jest.mock('../src/navigationStacks/AuthenticatedNavigator', () => {
  const {Text} = require('react-native');
  return ({userType, admin}: {userType: string; admin: string}) => {
    return (
      <Text
        testID={`auth-navigator-${userType}-${admin ? 'admin' : 'not-admin'}`}
      />
    );
  };
});

describe('App Component', () => {
  let store: Store;

  beforeEach(() => {
    store = createMockStore({
      auth: {isAuthenticated: false},
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly when authenticated', () => {
    const mockAuth = require('../src/features/auth/authSlice').authSlice;
    mockAuth.mockReturnValue({isAuthenticated: true});
    store = createMockStore({
      auth: {isAuthenticated: true},
    });

    const {toJSON} = render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('renders correctly when not authenticated', () => {
    const mockAuth = require('../src/features/auth/authSlice').authSlice;
    mockAuth.mockReturnValue({isAuthenticated: false});

    const {toJSON} = render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('useGetUserQuery is called with correct arguments when not authenticated', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const useGetUserQuery =
      require('../src/features/user/userApi').useGetUserQuery;
    expect(useGetUserQuery).toHaveBeenCalledWith(undefined, {
      skip: true,
      refetchOnMountOrArgChange: true,
    });
  });

  test('useGetUserQuery is called when authenticated', () => {
    const mockAuth = require('../src/features/auth/authSlice').authSlice;
    mockAuth.mockReturnValue({isAuthenticated: true});
    store = createMockStore({
      auth: {isAuthenticated: true},
    });

    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const useGetUserQuery =
      require('../src/features/user/userApi').useGetUserQuery;
    expect(useGetUserQuery).toHaveBeenCalledWith(undefined, {
      skip: false,
      refetchOnMountOrArgChange: true,
    });
  });

  test('useGetUserQuery is NOT called when not authenticated', () => {
    const mockAuth = require('../src/features/auth/authSlice').authSlice;
    mockAuth.mockReturnValue({isAuthenticated: false});

    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const useGetUserQuery =
      require('../src/features/user/userApi').useGetUserQuery;
    expect(useGetUserQuery).toHaveBeenCalledWith(undefined, {
      skip: true,
      refetchOnMountOrArgChange: true,
    });
  });

  test('useSignOutMutation is called when user signs out', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const useSignOutMutation =
      require('../src/features/auth/authApi').useSignOutMutation;
    expect(useSignOutMutation).toHaveBeenCalled();
  });

  test('renders GuestStackNavigator when not authenticated', () => {
    const {getByTestId} = render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    expect(getByTestId('guest-navigator')).toBeTruthy(); // Check if GuestNavigator is rendered
  });

  test('renders AuthenticatedNavigator with tenant userType', () => {
    const {getByTestId} = render(
      <AuthenticatedNavigator userType="tenant" admin={false} />,
    );

    expect(getByTestId('auth-navigator-tenant-not-admin')).toBeTruthy();
  });

  test('renders AuthenticatedNavigator with admin userType', () => {
    const {getByTestId} = render(
      <AuthenticatedNavigator userType="admin" admin={true} />,
    );

    expect(getByTestId('auth-navigator-admin-admin')).toBeTruthy();
  });
});
