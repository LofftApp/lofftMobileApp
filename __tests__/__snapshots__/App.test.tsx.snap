// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders correctly when authenticated 1`] = `
<Text
  testID="not-found-component"
>
  Error loading user type. Please try again
</Text>
`;

exports[`App Component renders correctly when not authenticated 1`] = `
<Text
  testID="guest-navigator"
/>
`;
